mkosi/__init__.py

```diff
@@ -3898,6 +3898,38 @@ def generate_btrfs(args: MkosiArgs, root: Path, label: str, for_cache: bool) ->
     return f
 
 
+def generate_xfs(args: MkosiArgs, root: Path, label: str, for_cache: bool) -> Optional[BinaryIO]:
+    if args.output_format != OutputFormat.gpt_xfs:
+        return None
+    if for_cache:
+        return None
+
+    @contextlib.contextmanager
+    def get_loopdev(f: Optional[BinaryIO]) -> Iterator[Optional[Path]]:
+        c = run(["losetup", "--find", "--show", f.name], stdout=PIPE, text=True)
+        l = Path(c.stdout.strip())
+        yield l
+        run(["losetup", "--detach", l])
+
+    with complete_step("Creating xfs root file system…"):
+        f: BinaryIO = cast(
+            BinaryIO,
+            tempfile.NamedTemporaryFile(prefix=".mkosi-mkfs-xfs", dir=os.path.dirname(args.output))
+        )
+
+        f.truncate(args.root_size)
+        run(mkfs_xfs_cmd(label) + [f.name])
+
+        with tempfile.TemporaryDirectory(prefix=".mkosi-xfs-mnt", dir=os.path.dirname(args.output)) as td:
+            with get_loopdev(f) as loopdev:
+                mount_point = Path(td)
+                mount_loop(args, loopdev, mount_point)
+                copy_path(root, mount_point)
+                umount(mount_point)
```

Comment:
Please use a XFS "protofile" instead, see the manpage: https://man7.org/linux/man-pages/man8/mkfs.xfs.8.html

Reply:
If I have this right I'd need to write a generator for a protofile based on the root directory and use this to load the file system contents. Is this what you're asking for? It seems the contents of such a file would certainly be quite big.

Reply:
Nvm, It seems I misunderstood the option. It doesn't do what I thought it did

Reply:
So what's next?

Reply:
I guess we'll have to do it like this? Ideally mkfs.xfs would support this via an option just like mkfs.btrfs and mkfs.ext4. I would like #1070 to land first with the changes to mount_loop()

Reply:
Please rebase and update the PR to use the new `mount_loop()` context manager

Reply:
Done.